improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,392 @@
+    ) -> None:
+        """Checks the size and units of input cubes"""
+
+        temp_slice = next(temperature_on_pressure.slices_over("pressure"))
+
+        assert_spatial_coords_match([ccl_temperature, ccl_pressure, temp_slice])
+        assert_spatial_coords_match(
+            [temperature_on_pressure, relative_humidity_on_pressure]
+        )
+
+        ccl_temperature.convert_units("K")
+        ccl_pressure.convert_units("Pa")
+        temperature_on_pressure.convert_units("K")
+        relative_humidity_on_pressure.convert_units("kg/kg")
+
+    def variable_at_pressure(
+        self, variable_on_pressure: Cube, pressure: np.ndarray
+    ) -> np.ndarray:
+        """Extracts the values from variable_on_pressure cube at pressure
+        levels described by the pressure array"""
+
+        variable = []
+        pressure_grid = self.pressure_grid(variable_on_pressure)
+
+        # deals with each realization independently and puts the arrays back together
+        for rea in range(len(pressure_grid)):
+            pressure_diff = abs(pressure_grid[rea] - pressure[rea])
+
+            indicies = np.nanargmin(pressure_diff, axis=0)
+            lat, long = indicies.shape
+            lat, long = np.ogrid[:lat, :long]
+
+            variable.append(variable_on_pressure.data[rea, indicies, lat, long])
+
+        return np.array(variable)
+
+    def pressure_grid(self, variable_on_pressure: Cube) -> np.ndarray:
+        """Creates a pressure_grid of the same shape as variable_on_pressure
+        with all possible pressure levels."""
+
+        required_shape = variable_on_pressure.shape
+        pressure_points = variable_on_pressure.coord("pressure").points
+        (pressure_axis,) = variable_on_pressure.coord_dims("pressure")
+        pressure_shape = np.ones_like(required_shape)
+        pressure_shape[pressure_axis] = required_shape[pressure_axis]
+        pressure_array = np.broadcast_to(
+            pressure_points.reshape(pressure_shape), required_shape
+        )
+        return pressure_array
+
+    def pressure_at_268(self, temperature_on_pressure: Cube) -> (CubeList):
+        """Extracts the lowest pressure level where the environment
+        temperature is below -5 degrees (268.15K). It also produces
+        the environemnt temperature at that pressure value"""
+
+        pressure_template = next(temperature_on_pressure.slices_over(["pressure"]))
+        pressure_template.rename("pressure of atmosphere at 268.15K")
+        pressure_template.units = temperature_on_pressure.coord("pressure").units
+        pressure_template.remove_coord("pressure")
+
+        temperature_template = next(temperature_on_pressure.slices_over(["pressure"]))
+        temperature_template.rename("tempeature_of_atmosphere_at_268.15K")
+        temperature_template.remove_coord("pressure")
+
+        shape = temperature_on_pressure.data.shape
+        r, l, m = (shape[0], shape[2], shape[3])  # ignores pressure axis
```

Comment:
🐰 🍋 🐒 
🦝 🥼 🥩 

I'm guessing at what r, l and m might mean as they aren't very descriptive. In fact they are so undescriptive I can't tell if we are dangerously assuming an order we haven't enforced again.

In the various cases I am highlighting where order assumptions are being made, these can all be accommodated using the `enforce_coordinate_ordering` function when you first load the data. This will give you confidence to make the assumptions you have. It comes with two additional considerations:
1. first you need to be sure the coordinates exist (see the comment about assuming there is a realization coord).
2. you need to return the data in the order the user expects. If they pass in a lon-lat grid and you reorder it internally, you need to ensure that you return a lon-lat grid at the end or they may be rather confused.

Reply:
You're absolutely right I was dangerously assuming an order. I've updated this and renamed the variable to hopefully be clearer what they are